refactor: Remove bok-choy and xblockutils (use xblock.utils instead) packages

farhan opened the pull request:
## Description

1. Removed xblock utils package and migrated to xblock package utils:
    -  Replace xblockutils.* imports with xblock.utils.* imports
    - ticket: https://github.com/openedx/xblock-drag-and-drop-v2/issues/356
2. Removed bok-choy packages:
    - Removed bok-choy package and its relevant test cases
    - ticket: https://github.com/openedx/xblock-drag-and-drop-v2/issues/354
3. Updated python requirements

## Breaking change in tests
There is no breaking change in the xblock feature in the implementation of this PR
As we have parted ways with the [bok-choy library](https://github.com/openedx/public-engineering/issues/13) this PR is removing all the integrating test cases along with the base test classes because they are linked with the library.

## Notes
Resolves #356 
Resolves #354

Agrendalath reviewed (changes requested):
@farhan, this should define a minimum version for the `XBlock` dependency in the `requirements/base.in`. 

We should bump the major version here if we do not provide the fallback import from `xblockutils`. We should also clearly state which of the edx-platform named releases are compatible with this new version.

Agrendalath commented:
@farhan, why are we removing integration tests? Are we going to replace them with something else?

farhan commented:
> @farhan, why are we removing integration tests? Are we going to replace them with something else?

@Agrendalath We are removing bok-choy integration tests because the library has been deprecated.

Bok-choy integration test has not been copied from [xblock-utils](https://github.com/openedx-unsupported/xblock-utils/tree/master/tests/integration) to [xblock/tests/utils](https://github.com/openedx/XBlock/tree/master/xblock/test/utils)

due to which the repositories that are using the base test classes of xblock/integration/tests are breaking.

We are already planning to remove bok-choy test cases, so covering it in this pr as well.
relevant ticket: https://github.com/openedx/xblock-drag-and-drop-v2/issues/354

salman2013 commented on tests/integration/test_events.py:
```diff
@@ -1,241 +0,0 @@
-
-from ddt import data, ddt, unpack
-from mock import Mock, patch
-from selenium.webdriver.common.keys import Keys
```
@feanil is it safe and fine to remove selenium along with bok-choy, apparently it looks like bok-choy uses selenium and there is no use of selenium without bok-choy but is that the case?

feanil replied:
Yes, selenium and bok-choy should be removed together.

salman2013 commented on requirements/test.in:
```diff
@@ -7,7 +7,6 @@ pytest-cov                # pytest extension for code coverage statistics
 ddt                       # data-driven tests
 
-bok-choy                  # integration tests
 selenium                  # integration tests
```
@farhan As you are removing selenium test cases can we also uninstall it from here?

farhan replied:
yes, i am removing, coming in next commit

salman2013 commented:
@farhan few test cases are failing like
https://github.com/openedx/xblock-drag-and-drop-v2/actions/runs/6545659095/job/17774598693?pr=357

farhan commented:
@Agrendalath 
I have completed my implementation.
I have removed all the things related to the integration test as all of them are linked to the Bok-choy test cases library.
I have removed the relevant configuration and docs related to integration tests as well. Please guide should we remove all of it as we don't have any plan in the pipeline to write integration tests again using other library.

> We should also clearly state which of the edx-platform named releases are compatible with this new version.
I have added the details in the change log, should we state this change on some other place as well?

Agrendalath commented:
@farhan

>  I have removed all the things related to the integration test as all of them are linked to the Bok-choy test cases library.
> I have removed the relevant configuration and docs related to integration tests as well. Please guide should we remove all of it as we don't have any plan in the pipeline to write integration tests again using other library.

What do you mean by "all of it"?

> I have added the details in the change log, should we state this change on some other place as well?

This is a breaking change, so we should indicate this [in the commit](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html), PR's name + description, and the changelog. Please take a look at #318 for an example of how we have been doing it.

farhan commented:
> What do you mean by "all of it"?

@Agrendalath I want to ask please verify all the removal done in this PR seems right to you

> Also, I wonder if xblock[django]==1.8.1 really is incompatible with older named releases. I don't see any breaking changes in its changelog or release notes.

That's true there is no breaking change in xblock release [v1.8.0](https://github.com/openedx/XBlock/releases/tag/xblock-1.8.0) and [v1.8.1](https://github.com/openedx/XBlock/releases/tag/xblock-1.8.1)

farhan commented:
@Agrendalath 
PR (Pull Request) is now ready for your review. I have also made necessary updates to the description, specifically addressing the breaking change. Your prompt review of these changes would be greatly appreciated.

Agrendalath commented on requirements/base.in:
```diff
@@ -3,4 +3,4 @@
 django-statici18n
 bleach[css]
-xblock-utils
+XBlock[django]>=1.8.1
```
If we have the `xblockutils` fallback, we can remove the constraint.

```suggestion
XBlock[django]
```

farhan replied:
I have a question here @Agrendalath 
what if someone is installing this xblock directly on the xblock-sdk and this repo is using a lesser version of xblock from v1.8.1. will it not break the xblock? (considering we have already removed xblock-utils package from this xblock)

Agrendalath replied:
@farhan, I see two possible cases:
1. Installing `xblock-sdk` from PyPI.
    In this scenario, dependencies are specified in`setup.py`, which [pulls them from `requirements/base.in`](https://github.com/openedx/xblock-sdk/blob/796c512/setup.py#L119) that contains [an unpinned version of `XBlock[django]`](https://github.com/openedx/xblock-sdk/blob/796c512/requirements/base.in#L14). Therefore, `pip` pulls the most recent version of `XBlock[django]`, which (at the moment) is `v1.8.1`.
1. Installing `xblock-sdk` dependencies from its requirements (e.g., after cloning the `xblock-sdk` repo for development).
    In this scenario, the developer uses [`make requirements` or `make install`](https://github.com/openedx/xblock-sdk/blob/796c512e5121639c4a15c3dc46659ff418630479/Makefile#L77-L81) _(side note: these recipes should use `pip-sync` to ensure that the removed packages are uninstalled from a virtualenv)_. In this scenario, [we already use the `xblock[django]==1.8.1`](https://github.com/openedx/xblock-sdk/blob/796c512/requirements/dev.txt#L378-L383).

Of course, the second scenario can install an older version if somebody does not use an up-to-date branch of `xblock-sdk`. However, it can be resolved with `git pull`, so it shouldn't be a big deal.

Agrendalath commented on requirements/workbench.in:
```diff
@@ -3,13 +3,10 @@
 
 -r test.txt               # Core and testing dependencies for this package
 
-mysqlclient
-
 fs-s3fs
 lxml
 requests
 pypng
 simplejson
 web-fragments
 webob
-XBlock[django]
```
The `workbench` dependencies were used only in integration tests. We can remove `workbench.in` and `workbench.txt` files. Then, we should remove their mentions in `Makefile` and `dev.in`.

Agrendalath commented on Makefile:
```diff
@@ -62,36 +60,26 @@ push_translations: ## push translations to transifex
 
-test.python: ## run python unit and integration tests
-	PATH=test_helpers/firefox:$$PATH xvfb-run python run_tests.py $(TEST)
+test.python: ## run python unit tests
+	python run_tests.py $(TEST)
```
```suggestion
test.python: ## run python unit tests in the local virtualenv
	pytest --cov drag_and_drop_v2 $(TEST)
```

Then, let's remove `run_tests.py` and replace the example in the README with `make test.python TEST=tests/unit/test_basics.py::BasicTests::test_student_view_data`.

Agrendalath commented on Changelog.md:
```diff
@@ -4,6 +4,14 @@ Drag and Drop XBlock changelog
+
+* Removed xblock-utils package
+  * Replace xblockutils.* imports with xblock.utils.* imports
+* Removed bok-choy package and its relevant test cases.
```
```suggestion
* Removed bok-choy package along with all integration tests.
```

farhan commented:
@Agrendalath Addressed all the requested changes.
PR is available for next pass

feanil commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -23,9 +23,9 @@
 from xblock.fields import Boolean, Dict, Float, Integer, Scope, String
 from xblock.scorable import ScorableXBlockMixin, Score
+from xblock.utils.resources import ResourceLoader
+from xblock.utils.settings import ThemableXBlockMixin, XBlockWithSettingsMixin
```
Yea, I think we should do this for other xblocks as well to maintain compatibility for as long as possible since it's not much more work to do it that way.
